src/lib/operators/insert.cpp

```diff
@@ -106,13 +106,18 @@ std::shared_ptr<const Table> Insert::_on_execute(std::shared_ptr<TransactionCont
     if (_target_table->chunk_count() == 0) {
       _target_table->append_mutable_chunk();
     }
+    const auto chunks_before_insert = _target_table->chunk_count();
 
     while (remaining_rows > 0) {
       auto target_chunk_id = ChunkID{_target_table->chunk_count() - 1};
       auto target_chunk = _target_table->get_chunk(target_chunk_id);
 
       // If the last Chunk of the target Table is either immutable or full, append a new mutable Chunk
       if (!target_chunk->is_mutable() || target_chunk->size() == _target_table->max_chunk_size()) {
+        // Finalize the last Chunk only if it was not added by us.
+        if (target_chunk->is_mutable() && chunks_before_insert - 1 == target_chunk_id)
+          target_chunk->finalize();
```

Comment:
Wait a second - how can we be sure that everyone who is writing on this table has set their begin CIDs?

Reply:
Have a look at `ChunkCompressionTask::_chunk_is_completed`. Maybe we can (1) merge those and (2) make sure that once the last insert tasks finishes or aborts, it is responsible for finalizing the chunk. Good potential for race conditions though.

Also, this shows that we probably have a missing test.

Reply:
The content of `_chunk_is_completed` makes sense, but how do we find out which Insert is the last? A solution would be a periodic task that checks for chunks that reached their capacity before finalizing and encoding them.

Reply:
I tried to come up with a way where we identify the last writing transaction, but it will be hard to prove thread safety for those. As of now, I would consider a separate task and doing this asynchronously more intuitive. Maybe this could be merged with the MVCC cleanup?

Reply:
Let's discuss that offline.

Reply:
We decided to remove the finalization from the Insert operator for this PR. The validation optimization can be used without it, too. It might not be applied for as many cases as it would be with the finalization in Insert. The code and test were moved to bensk1/insert_finalization

Reply:
Could you add a summary of this discussion/outcome to the description?

I can’t follow.